### PR TITLE
Adding CRDs description from Design.md

### PR DESCRIPTION
**Pull request by AshwinSriram11**
## Description

According to the discussion https://github.com/prometheus-operator/prometheus-operator/pull/6628#discussion_r1624682243, I have added the description of each CRD in the Design.md page to appropriate types.go files in pkg directory and api.md page for each package. This PR only contains changes to Prometheus, Alertmanager, ThanosRuler and PrometheusAgent

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [x] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

### CODE REVIEW

**Comment by AshwinSriram11** [comment]
I did not add the examples mentioned for PodMonitor, ServiceMonitor and AlertmanagerConfig to the description. I have also kept the Design.md doc as it is for now. Maybe we can move those examples somewhere else and then remove Design.md?

File: pkg/apis/monitoring/v1/alertmanager_types.go
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/alertmanager_types.go:41]:
```suggestion
// The `Alertmanager` custom resource definition (CRD) defines a desired [Alertmanager](https://prometheus.io/docs/alerting) setup to run in a Kubernetes cluster. It allows to specify many options such as the number of replicas, persistent storage and many more.
```

File: pkg/apis/monitoring/v1/alertmanager_types.go
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/alertmanager_types.go:45]: (suggestion) the secret part is a bit outdated since you can also use AlertmanagerConfig objects.

```suggestion
// For each `Alertmanager` resource, the Operator deploys a `StatefulSet` object in the same namespace. When there are two (or more) replicas, it configures automatically the Alertmanager pods in high-availability mode.
```

**Review by simonpasquier — COMMENTED** [review]
I only reviewed the Alertmanager description. It's also a good occasion to clean up/simplify the wording.
Maybe split between pod-based resources (Alertmanager, Prometheus, ...) and metrics-based resources (ServiceMonitor, PodMonitor, ...) to speed up the reviews?

File: pkg/apis/monitoring/v1/podmonitor_types.go
Comment (ArthurSens) [inline pkg/apis/monitoring/v1/podmonitor_types.go:38]:
```suggestion
```

I think we don't need to mention what a pod is :)

File: pkg/apis/monitoring/v1/podmonitor_types.go
Comment (ArthurSens) [inline pkg/apis/monitoring/v1/podmonitor_types.go:40]:
```suggestion
// The `PodMonitor` object introduced by the Prometheus Operator discovers pods and generates the relevant configuration for the Prometheus server to monitor them.
```

File: pkg/apis/monitoring/v1/podmonitor_types.go
Comment (ArthurSens) [inline pkg/apis/monitoring/v1/podmonitor_types.go:47]: Like this how?

Should this line be deleted?

File: pkg/apis/monitoring/v1/thanos_types.go
Comment (ArthurSens) [inline pkg/apis/monitoring/v1/thanos_types.go:45]: Do we know if this link will continue to work when moving things around?

**Review by ArthurSens — COMMENTED** [review]
We could also improve the documentation of `ScrapeConfig`, in a separate PR of course :)

File: pkg/apis/monitoring/v1/podmonitor_types.go
Comment (AshwinSriram11) [reply pkg/apis/monitoring/v1/podmonitor_types.go]: I will fix this in the next PR.

File: pkg/apis/monitoring/v1/thanos_types.go
Comment (AshwinSriram11) [reply pkg/apis/monitoring/v1/thanos_types.go:45]: I think the link on line 45 will not work because originally this section was in Design.md and Thanos.md were in the same directory. I will look into this.

File: pkg/apis/monitoring/v1/prometheus_types.go
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:748]:
```suggestion
// The `Prometheus` custom resource definition (CRD) defines a desired [Prometheus](https://prometheus.io/docs/prometheus) setup to run in a Kubernetes cluster. It allows to specify many options such as the number of replicas, persistent storage, Alertmanagers where firing alerts should be sent and many more.
```

File: pkg/apis/monitoring/v1/prometheus_types.go
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:750]:
```suggestion
// For each `Prometheus` resource, the Operator deploys one or several `StatefulSet` objects in the same namespace. The number of StatefulSets is equal to the number of shards which is 1 by default.
```

File: pkg/apis/monitoring/v1/alertmanager_types.go
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/alertmanager_types.go:43]: To be consistent with Prometheus, we could talk about AlertmanagerConfig selection (a follow-up PR is fine).

File: pkg/apis/monitoring/v1/prometheus_types.go
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:752]:
```suggestion
// The resource defines via label and namespace selectors which `ServiceMonitor`, `PodMonitor`, `Probe`  and `PrometheusRule` objects should be associated to the deployed Prometheus instances.
//
// The Operator continuously reconciles the scrape and rules configuration and a sidecar container running in the Prometheus pods triggers a reload of the configuration when needed.
```

File: pkg/apis/monitoring/v1/thanos_types.go
Comment (simonpasquier) [reply pkg/apis/monitoring/v1/thanos_types.go:45]: Yes I would remove the link to avoid trouble.

File: pkg/apis/monitoring/v1/alertmanager_types.go
Comment (AshwinSriram11) [reply pkg/apis/monitoring/v1/alertmanager_types.go:43]: Maybe something like

> The resource defines via label and namespace selectors which `AlertmanagerConfig` objects should be associated to the deployed Alertmanger instances.

Comment (simonpasquier) [reply pkg/apis/monitoring/v1/alertmanager_types.go:43]: yes

File: pkg/apis/monitoring/v1/thanos_types.go
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/thanos_types.go:43]: (nit)
```suggestion
// A `ThanosRuler` instance requires at least one compatible Prometheus API endpoint (either Thanos Querier or Prometheus services).
```

File: pkg/apis/monitoring/v1/thanos_types.go
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/thanos_types.go:41]:
```suggestion
// The `ThanosRuler` custom resource definition (CRD) defines a desired [Thanos Ruler](https://github.com/thanos-io/thanos/blob/main/docs/components/rule.md) setup to run in a Kubernetes cluster.
```

File: pkg/apis/monitoring/v1alpha1/prometheusagent_types.go
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/prometheusagent_types.go:56]:
```suggestion
// The `PrometheusAgent` custom resource definition (CRD) defines a desired [Prometheus Agent](https://prometheus.io/blog/2021/11/16/agent/) setup to run in a Kubernetes cluster.
```

File: pkg/apis/monitoring/v1alpha1/prometheusagent_types.go
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/prometheusagent_types.go:58]:
```suggestion
// The CRD is very similar to the `Prometheus` CRD except for features which aren't available in agent mode like rule evaluation, persistent storage and Thanos sidecar.
```

File: pkg/apis/monitoring/v1alpha1/prometheusagent_types.go
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/prometheusagent_types.go:60]: (suggestion) I would omit this link in the API documentation.
```suggestion
```

File: pkg/apis/monitoring/v1/thanos_types.go
Comment (ArthurSens) [inline pkg/apis/monitoring/v1/thanos_types.go:43]: For consistency with Prometheus/Alertmanager, we can also mention the PrometheusRule selection for thanos ruler

```
// The resource defines via label and namespace selectors which `PrometheusRule` objects should be associated with the deployed Thanos Ruler instances.
```

**Review by ArthurSens — DISMISSED** [review]
But it does LGTM, I'm giving the approval already and I'm fining merging it.

I'd just wait for Simon's approval as well or if you'd like to wait for @nicolastakashi's input

File: pkg/apis/monitoring/v1/thanos_types.go
Comment (AshwinSriram11) [reply pkg/apis/monitoring/v1/thanos_types.go:43]: Simon and Nicolas might have some suggestions, so let's wait for them and then I will add this together with it.

Comment (simonpasquier) [reply pkg/apis/monitoring/v1/thanos_types.go:43]: Agree with @ArthurSens. Something along these lines?

```
The resource defines via label and namespace selectors 
which `PrometheusRule` objects should be associated to the 
deployed Thanos Ruler instances.
````

**Review by simonpasquier — APPROVED** [review]
thanks!

**Comment by simonpasquier** [comment]
I'm fine with it too, thanks!

**Pull request merged**
